### PR TITLE
Select humanoid tracking example via button

### PR DESCRIPTION
**Pull request by thowell**
Logic to select motion that is tracked in humanoid tracking task. Utilizes `userdata` to store the current motion id in `userdata[0]` and the simulation time in `userdata[1]`. A future PR (after #65 is incorporated) will enabling looping through all tasks.
@hartikainen

### CODE REVIEW

File: mjpc/tasks/humanoid/tracking/task.xml [hunk 1]
```diff
@@ -137,14 +138,14 @@
     <framelinvel name="tracking_linvel[rhip]" objtype="site" objname="tracking[rhip]" />
   </sensor>
 
-  <!-- <include file="./keyframes/CMU-CMU-02-02_04_poses.xml" />  simple jump, balance -->
-  <!-- <include file="./keyframes/CMU-CMU-87-87_01_poses.xml" />  <!-\- jump with kick and spin -\-> -->
-  <!-- <include file="./keyframes/CMU-CMU-88-88_06_poses.xml" />  <!-\- jump and spin kick -\-> -->
-  <!-- <include file="./keyframes/CMU-CMU-88-88_07_poses.xml" />  <!-\- cartwheel-\-> -->
-  <!-- <include file="./keyframes/CMU-CMU-88-88_08_poses.xml" />  <!-\- crouch and flip backward on hands -\-> -->
-  <!-- <include file="./keyframes/CMU-CMU-88-88_09_poses.xml" />  <!-\- stretch and cartwheel -\-> -->
+  <include file="./keyframes/CMU-CMU-02-02_04_poses.xml" />  simple jump, balance
```
Comment (hartikainen) [inline mjpc/tasks/humanoid/tracking/task.xml:141]: The comment on this line is different from the ones below. Maybe change to `<!-\- simple jump, balance -\->` or remove `<!-\-` and `-\->` from the others

File: mjpc/tasks/humanoid/tracking/task.cc [hunk 2]
```diff
@@ -191,10 +243,35 @@ void humanoid::Tracking::Residual(const double *parameters,
...
+    d->userdata[1] = d->time;  // reference time
+
+    // set home state
+    double* qpos_home = model->key_qpos + model->nq * start;
+    double* qvel_home = model->key_qvel + model->nv * start;
+
+    mju_copy(d->qpos, qpos_home, model->nq);
+    mju_copy(d->qvel, qvel_home, model->nv);
+
+    printf("Motion Switch\n");
```
Comment (hartikainen) [inline mjpc/tasks/humanoid/tracking/task.cc:269]: Maybe print the motion index/name here too.

File: mjpc/tasks/humanoid/tracking/task.cc [hunk 3]
```diff
@@ -191,10 +243,35 @@ void humanoid::Tracking::Residual(const double *parameters,
...
+  }
+
+  // indices
+  double current_index = (d->time - d->userdata[1]) * fps + start;  
+  int last_key_index = start + length - 1; // TODO(taylor): check...
```
Comment (hartikainen) [inline mjpc/tasks/humanoid/tracking/task.cc:274]: This seems correct to me (re your `TODO` here).

File: mjpc/tasks/humanoid/tracking/task.cc [hunk 4]
```diff
@@ -37,6 +38,55 @@ std::tuple<int, int, double, double> ComputeInterpolationValues(double index,
...
+    return 279;
+  // Run
+  } else if (id == 8) {
+    return 39;
+  // Walk
+  } else if (id == 9) {
+    return 510;
+  // Stand
+  } else if (id == 10) {
+    return 1;
```
Comment (hartikainen) [inline mjpc/tasks/humanoid/tracking/task.cc:76]: Is this one missing the actual keyframe?

File: mjpc/tasks/humanoid/tracking/task.cc [hunk 5]
```diff
@@ -37,6 +38,55 @@ std::tuple<int, int, double, double> ComputeInterpolationValues(double index,
 
   return {index_0, index_1, weight_0, weight_1};
 }
+
+// return length of motion trajectory
+int TrajectoryLength(int id) {
+  // Jump
+  if (id == 0) {
```
Comment (hartikainen) [inline mjpc/tasks/humanoid/tracking/task.cc:45]: Can we comment this also with the the mocap filename ids (e.g. `CMU-CMU-88-88_09_poses.xml`).

**Review by hartikainen — COMMENTED** [review]
Looks good to me!

File: mjpc/tasks/humanoid/tracking/task.cc [hunk 6]
```diff
@@ -191,10 +243,35 @@ void humanoid::Tracking::Residual(const double *parameters,
...
+  if ((int)d->userdata[0] != task->stage || d->time == 0.0) {
+    d->userdata[0] = task->stage; // set motion id
+    d->userdata[1] = d->time;  // reference time
+
+    // set home state
+    double* qpos_home = model->key_qpos + model->nq * start;
+    double* qvel_home = model->key_qvel + model->nv * start;
+
+    mju_copy(d->qpos, qpos_home, model->nq);
+    mju_copy(d->qvel, qvel_home, model->nv);
```
Comment (hartikainen) [inline mjpc/tasks/humanoid/tracking/task.cc:267]: Maybe set with `mj_resetDataKeyframe`?

File: mjpc/tasks/humanoid/tracking/task.cc (hunk 4 above)
Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.cc:76]: Removed.

File: mjpc/tasks/humanoid/tracking/task.cc (hunk 3 above)
Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.cc:274]: Comment removed.

File: mjpc/tasks/humanoid/tracking/task.cc (hunk 2 above)
Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.cc:269]: Done.

File: mjpc/tasks/humanoid/tracking/task.cc (hunk 5 above)
Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.cc:45]: Done.

File: mjpc/tasks/humanoid/tracking/task.xml (hunk 1 above)
Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.xml:141]: Standardized the comment style.

File: mjpc/tasks/humanoid/tracking/task.cc (hunk 6 above)
Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.cc:267]: Done.

**Comment by thowell** [comment]
@hartikainen thanks for you comments. I addressed each one and made most of the changes that you suggested. 
@yuvaltassa can you review and merge?

**Comment by yuvaltassa** [comment]
Give me a day or two?, @erez-tom's restructure is nearly ready to submit.

**Comment by thowell** [comment]
Updated to merge in latest changes to main. Additionally, updated Quadruped Hill and Quadrotor tasks to use `Task` member `int current_stage` instead of `userdata[0]`.

File: mjpc/tasks/humanoid/tracking/task.xml [hunk 7]
```diff
@@ -1,9 +1,10 @@
 <mujoco model="Humanoid Track">
   <include file="../../common.xml"/>
   <include file="../humanoid.xml" />
-  <size memory="400K"/>
+  <size memory="400K" nuserdata="2"/> <!-- userdata: [motion id, reference time]-->
```
Comment (erez-tom) [inline mjpc/tasks/humanoid/tracking/task.xml:4]: you probably don't need nuserdata anymore, right?  maybe also in the quadrotor?

Comment (thowell) [reply mjpc/tasks/humanoid/tracking/task.xml]: Removed from all task xml files now.

**Review by erez-tom — APPROVED** [review]

**Review by erez-tom — APPROVED** [review]

**Pull request merged**